File: backend/app/utils/deep_links.py
"""
VistaTrek Deep Links
Navigation app deep link generation
"""

from urllib.parse import quote

from app.models.schemas import Coordinates, NavApp


def generate_nav_link(
    destination: Coordinates,
    app: NavApp,
    destination_name: str = "",
) -> str:
    """
    Generate deep link for navigation app.

    Args:
        destination: Target coordinates
        app: Navigation app (waze, google, apple)
        destination_name: Optional name for the destination

    Returns:
        Deep link URL string
    """
    lat, lon = destination.lat, destination.lon
    name_encoded = quote(destination_name) if destination_name else ""

    if app == NavApp.WAZE:
        # Waze deep link
        return f"https://waze.com/ul?ll={lat},{lon}&navigate=yes"

    elif app == NavApp.GOOGLE:
        # Google Maps deep link
        if destination_name:
            return f"https://www.google.com/maps/dir/?api=1&destination={lat},{lon}&destination_place_id={name_encoded}&travelmode=driving"
        return f"https://www.google.com/maps/dir/?api=1&destination={lat},{lon}&travelmode=driving"

    elif app == NavApp.APPLE:
        # Apple Maps deep link
        if destination_name:
            return f"https://maps.apple.com/?daddr={lat},{lon}&dirflg=d&t=m&q={name_encoded}"
        return f"https://maps.apple.com/?daddr={lat},{lon}&dirflg=d&t=m"

    else:
        # Default to Google Maps
        return f"https://www.google.com/maps/dir/?api=1&destination={lat},{lon}&travelmode=driving"


def generate_nav_link_with_waypoints(
    origin: Coordinates,
    destination: Coordinates,
    waypoints: list[Coordinates],
    app: NavApp,
) -> str:
    """
    Generate navigation link with multiple waypoints.
    Note: Not all apps support waypoints equally well.

    Args:
        origin: Starting point
        destination: End point
        waypoints: Intermediate stops
        app: Navigation app

    Returns:
        Deep link URL string
    """
    if app == NavApp.GOOGLE:
        # Google Maps supports waypoints
        waypoints_str = "|".join(f"{wp.lat},{wp.lon}" for wp in waypoints)
        return (
            f"https://www.google.com/maps/dir/?api=1"
            f"&origin={origin.lat},{origin.lon}"
            f"&destination={destination.lat},{destination.lon}"
            f"&waypoints={waypoints_str}"
            f"&travelmode=driving"
        )

    elif app == NavApp.WAZE:
        # Waze doesn't support waypoints in deep links
        # Navigate to first waypoint, user will need to continue manually
        if waypoints:
            return generate_nav_link(waypoints[0], NavApp.WAZE)
        return generate_nav_link(destination, NavApp.WAZE)

    elif app == NavApp.APPLE:
        # Apple Maps has limited waypoint support
        # Use the first waypoint as a via point
        if waypoints:
            return (
                f"https://maps.apple.com/?saddr={origin.lat},{origin.lon}"
                f"&daddr={waypoints[0].lat},{waypoints[0].lon}+to:{destination.lat},{destination.lon}"
                f"&dirflg=d"
            )
        return (
            f"https://maps.apple.com/?saddr={origin.lat},{origin.lon}"
            f"&daddr={destination.lat},{destination.lon}"
            f"&dirflg=d"
        )

    return generate_nav_link(destination, app)

File: frontend/src/utils/deepLinks.ts
/**
 * Deep Links Generator
 * Per PRD: Generate navigation app links (Waze, Google Maps, Apple Maps)
 */

import { Coordinates, NavApp } from '../types';

/**
 * Generate a navigation deep link for the given app
 */
export function generateNavLink(
  destination: Coordinates,
  app: NavApp,
  name?: string
): string {
  const { lat, lon } = destination;
  const encodedName = name ? encodeURIComponent(name) : '';

  switch (app) {
    case 'waze':
      // Waze deep link format
      return `https://waze.com/ul?ll=${lat},${lon}&navigate=yes`;

    case 'google':
      // Google Maps deep link format
      const googleParams = new URLSearchParams({
        api: '1',
        destination: `${lat},${lon}`,
      });
      if (name) {
        googleParams.set('destination_place_id', encodedName);
      }
      return `https://www.google.com/maps/dir/?${googleParams.toString()}`;

    case 'apple':
      // Apple Maps deep link format
      const appleParams = new URLSearchParams({
        daddr: `${lat},${lon}`,
        dirflg: 'd', // driving
      });
      if (name) {
        appleParams.set('q', name);
      }
      return `https://maps.apple.com/?${appleParams.toString()}`;

    default:
      // Fallback to Google Maps
      return `https://www.google.com/maps/dir/?api=1&destination=${lat},${lon}`;
  }
}

/**
 * Generate a navigation link with waypoints
 * Note: Only Google Maps supports waypoints in deep links
 */
export function generateNavLinkWithWaypoints(
  origin: Coordinates,
  destination: Coordinates,
  waypoints: Coordinates[],
  app: NavApp,
  destinationName?: string
): string {
  if (waypoints.length === 0) {
    return generateNavLink(destination, app, destinationName);
  }

  // Only Google Maps supports waypoints
  if (app === 'google') {
    const waypointsStr = waypoints
      .map((wp) => `${wp.lat},${wp.lon}`)
      .join('|');

    const params = new URLSearchParams({
      api: '1',
      origin: `${origin.lat},${origin.lon}`,
      destination: `${destination.lat},${destination.lon}`,
      waypoints: waypointsStr,
      travelmode: 'driving',
    });

    return `https://www.google.com/maps/dir/?${params.toString()}`;
  }

  // For other apps, just navigate to destination
  return generateNavLink(destination, app, destinationName);
}

/**
 * Open navigation in the appropriate app
 */
export function openNavigation(
  destination: Coordinates,
  app: NavApp,
  name?: string
): void {
  const url = generateNavLink(destination, app, name);
  window.open(url, '_blank');
}

/**
 * Check if the device likely supports a specific nav app
 */
export function isNavAppAvailable(app: NavApp): boolean {
  const userAgent = navigator.userAgent.toLowerCase();
  const isIOS = /iphone|ipad|ipod/.test(userAgent);
  const isAndroid = /android/.test(userAgent);

  switch (app) {
    case 'apple':
      return isIOS;
    case 'waze':
    case 'google':
      return isIOS || isAndroid;
    default:
      return true;
  }
}
